src/sage/combinat/posets/linear_extensions.py

```diff
@@ -252,6 +252,73 @@ def is_greedy(self):
                         return False
         return True
 
+    def is_supergreedy(self):
+        r"""
+        Return ``True`` if the linear extension is supergreedy.
+        
+        A linear extension `[x_1<x_2<...<x_t]` of a finite ordered
+        set `P=(P, <)` is *super greedy* if it can be obtained using
+        the following procedure: Choose `x_1` to be a minimal
+        element of `P`; suppose `x_1,...,x_i` have been chosen;
+        define `p(x)` to be the largest `j\leq i` such that `x_j<x`
+        if such a `j` exists and 0 otherwise; choose `x_{i+1}`
+        to be a minimal element of `P-\{x_1,...,x_i\}` which
+        maximizes `p`.
+
+        Informally, a linear extension is supergreedy if it "always
+        goes up and receedes the least"; in other words, supergreedy
+        linear extensions are depth-first linear extensions.
+        For more details see [KTZ1987]_.
+
+        EXAMPLES::
+        
+            sage: X = [0,1,2,3,4,5,6]
+            sage: Y = [[0,5],[1,4],[1,5],[3,6],[4,3],[5,6],[6,2]]
+            sage: P = Poset((X,Y), cover_relations = True, facade=False)
+            sage: for l in P.linear_extensions():
+            ....:     if l.is_supergreedy():
+            ....:         print(l)
+            [1, 4, 3, 0, 5, 6, 2]
+            [0, 1, 4, 3, 5, 6, 2]
+            [0, 1, 5, 4, 3, 6, 2]
+
+            sage: Q = posets.PentagonPoset()
+            sage: for l in Q.linear_extensions():
+            ....:     if not l.is_supergreedy():
+            ....:         print(l)
+            [0, 2, 1, 3, 4]
+
+        TESTS::
+
+            sage: T = Poset()
+            sage: T.linear_extensions()[0].is_supergreedy()
+            True
+        """
+        P = self.poset()
+        H = P.hasse_diagram()
+
+        def next_elements(H, linext):
```

Comment:
`H` is not modified, why to you pass it as an argument to `next_elements()` ? 

Besides, I think that  a nested function is only needed when you do recursive calls of it, and/or
if you call it more than in one place in the code. Here you can just put its body where your line 315 
(i.e. where it is called).